wasm-thumbnail-rb/lib/wasm/thumbnail/rb.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require_relative "rb/version"
+require "wasmer"
+module Wasm
+  module Thumbnail
+    # As opposed to the PY implementation
+    module Rb
+      class Error < StandardError; end
+
+      # Now the module is compiled, we can instantiate it. Doing so outside the method where used results in errors.
+      def self.register_panic(_msg_ptr = nil, _msg_len = nil, _file_ptr = nil, _file_len = nil, _line = nil, _column = nil)
+        puts("WASM panicked")
+      end
+
+      WASMStore = Wasmer::Store.new
+      WASMImportObject = Wasmer::ImportObject.new
+      WASMImportObject.register(
+        "env",
+        register_panic: Wasmer::Function.new(
+          WASMStore,
+          :register_panic,
+          Wasmer::FunctionType.new([Wasmer::Type::I32,
+                                    Wasmer::Type::I32,
+                                    Wasmer::Type::I32,
+                                    Wasmer::Type::I32,
+                                    Wasmer::Type::I32,
+                                    Wasmer::Type::I32], [])
+        )
+      )
+
+      def self.resize_and_pad(file_bytes:, width:, height:, size:)
+        # Let's compile the module to be able to execute it!
+        wasm_instance = Wasm::Thumbnail::Rb::GetWasmInstance.call
+
+        # This tells us how much space we'll need to put our image in the WASM env
+        image_length = file_bytes.length
+        input_pointer = wasm_instance.exports.allocate.call(image_length)
+        # Get a pointer on the allocated memory so we can write to it
+        memory = wasm_instance.exports.memory.uint8_view input_pointer
+
+        # Put the image to resize in the allocated space
+        (0..image_length - 1).each do |nth|
+          memory[nth] = file_bytes[nth]
+        end
+
+        # Do the actual resize and pad
+        # Note that this writes to a portion of memory the new JPEG file, but right pads the rest of the space
+        # we gave it with 0.
+        begin
+          output_pointer = wasm_instance.exports.resize_and_pad.call(input_pointer,
+                                                                     image_length,
+                                                                     width,
+                                                                     height,
+                                                                     size)
+        rescue RuntimeError
+          raise "Error processing the image."
+        end
+        # Get a pointer to the result
+        memory = wasm_instance.exports.memory.uint8_view output_pointer
+
+        # Only take the buffer that we told the rust function we needed. The resize function
+        # makes a smaller image than the buffer we said, and then pads out the rest so we have to
+        # go hunting for the bytes which represent the JPEG image. In hex, JPEG images start with
+        # FFD8 and FFD9, so we can convert to hex and find the bounds of the image, then write to file
```

Comment:
we can drop this comment about hunting for magic bytes - right?

Reply:
Done.